tools/runner.py

```diff
@@ -0,0 +1,181 @@
+            action1 = player1.get_next_action(state)
+            state1, reward1, done1, _ = env.step(action1)
+
+            if player2_learn:
+                if LEARNING:
+                    player2.learn(state1, action2, env._reverse_reward(reward1), state, done1)
+            else:
+                player2_learn = True
+
+            if not done1:
+                action2 = player2.get_next_action(state1)
+                state2, reward2, done2, _ = env.step(action2)
+
+                if LEARNING:
+                    player1.learn(state, action1, env._reverse_reward(reward2), state2, done2)
+
+                if done2:
+                    done = True
+                    if LEARNING:
+                        player2.learn(state1, action2, reward2, state2, done2)
+
+                    if reward2 != env.DRAW_REWARD:
+                        # player2 Won
+                        match_result = -1
+                    else:
+                        # player2 Draw
+                        match_result = 0
+
+                state = state2
+            else:
+                done = True
+                if LEARNING:
+                    player1.learn(state, action1, reward1, state1, done1)
+
+                if reward1 != env.DRAW_REWARD:
+                    # player1 Won
+                    match_result = 1
+                else:
+                    # player1 Draw
+                    match_result = 0
+
+        # idx0 = Win, idx1 = Draw, idx2 = Loss
+        results[1 - match_result] += 1
+        print(f"{player1.name}:{player2.name}={results}")
+
+    return results
+
+
+def tournament_print(leaderboard):
+    for item in leaderboard:
+        leaderboard[item][4] = tournament_score(leaderboard[item])
+
+    leaderboard = sorted(leaderboard.items(), reverse=True, key=lambda itm: itm[1][4])
+
+    print("{:6s} {:25s} {:7s} | {:23s} | {:23s} | {:23s} | {:23s}".format("", "", "", "Starter Rounds", "Second Rounds", "Starter matches", "Second matches"))
+
+    print("{:6s} {:25s} {:7s} | {:7s} {:7s} {:7s} | {:7s} {:7s} {:7s} | {:7s} {:7s} {:7s} | {:7s} {:7s} {:7s}".format("Place", "Name", "Score", "Wins", "Draws", "Losses", "Wins", "Draws", "Losses", "Wins", "Draws", "Losses", "Wins", "Draws", "Losses"))
+    idx = 0
+    for item in leaderboard:
+        idx += 1
+        print(
+            "{:6d} {:25s} {:7d} | {:7d} {:7d} {:7d} | {:7d} {:7d} {:7d} | {:7d} {:7d} {:7d} | {:7d} {:7d} {:7d}".
+                format(idx, item[0], item[1][4],
+                       item[1][0][0], item[1][0][1], item[1][0][2],
+                       item[1][1][0], item[1][1][1], item[1][1][2],
+                       item[1][2][0], item[1][2][1], item[1][2][2],
+                       item[1][3][0], item[1][3][1], item[1][3][2]
+                       ))
+    pass
+
+
+def tournament_score(score):
+    # Score = Number of wins -  Number of losses
+    return (score[0][0] - score[0][2]) + (score[1][0] - score[1][2])
+
+
+def tournament(models):
```

Comment:
This function contains some obfuscated code, quite unclear on what it does. If we want it to use for scoring the players in final competition, we'd need to cover it at least by some unit-tests.